File: 15. 3Sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res=[]
        for i in range(len(nums)-2):            
            if i>0 and nums[i]==nums[i-1]:
                continue
            l=i+1
            r=len(nums)-1
            
            while l<r:
                sumi=nums[i]+nums[l]+nums[r]
                if sumi==0:
                    res.append([nums[i],nums[l],nums[r]])
                    while l<r and nums[l]==nums[l+1]:
                        l+=1
                    l+=1
                elif sumi>0:
                    r-=1
                else:
                    l+=1                                      
        return res
                               
                    
                

File: 15. 3Sum.cpp
class Solution {
public:
  vector<vector<int>> threeSum(vector<int> &nums) {
    int l, r, total;
    vector<vector<int>> res;
    vector<int> tmp;
    sort(nums.begin(), nums.end());
    for (int i; i < nums.size() - 2; i++) {
      if (i > 0 && nums[i - 1] == nums[i])
        continue;
      l = i + 1;
      r = nums.size() - 1;
      while (l < r) {
        total = nums[i] + nums[r] + nums[l];
        if (total == 0) {
          res.push_back(vector<int>{nums[i], nums[r], nums[l]});
          while (l < r && nums[l] == nums[l + 1])
            ++l;
          ++l;
        } else if (total > 0) {
          r--;
        } else {
          l++;
        }
      }
    }

    return res;
  }
};
